### PR TITLE
GNN ensembling for node-level prediction

### PR DESCRIPTION
**Pull request by weihua916**
Add model ensembling option

### CODE REVIEW

**Comment by weihua916** [comment]
rel-f1-position
```
Testing for ensemble indices from 0 to 0
Best test metrics: {'mae': 4.306731792332833, 'rmse': 5.258358217732742}
Testing for ensemble indices from 0 to 1
Best test metrics: {'mae': 4.2294558866400465, 'rmse': 5.180398734042701}
Testing for ensemble indices from 0 to 2
Best test metrics: {'mae': 4.221995389712484, 'rmse': 5.125791770671216}
Testing for ensemble indices from 0 to 3
Best test metrics: {'mae': 4.170969367905666, 'rmse': 5.037310977982346}
Testing for ensemble indices from 0 to 4
Best test metrics: {'mae': 4.132560997636695, 'rmse': 4.989645639655038}
Testing for ensemble indices from 0 to 5
Best test metrics: {'mae': 4.110166810897359, 'rmse': 4.979252859427289}
Testing for ensemble indices from 0 to 6
Best test metrics: {'mae': 4.104200592793917, 'rmse': 4.98705414124849}
Testing for ensemble indices from 0 to 7
Best test metrics: {'mae': 4.073461746123799, 'rmse': 4.941380638320971}
Testing for ensemble indices from 0 to 8
Best test metrics: {'mae': 4.043316612954726, 'rmse': 4.908553420247133}
Testing for ensemble indices from 0 to 9
Best test metrics: {'mae': 4.024014233497152, 'rmse': 4.8827221566276515}
```

**Comment by weihua916** [comment]
`rel-trial-adverse`
44.24 MAE -> 43.39 MAE
```
Testing for ensemble indices from 0 to 0
Best test metrics: {'mae': 44.24175917534693, 'rmse': 226.31972222562877}
Testing for ensemble indices from 0 to 1
Best test metrics: {'mae': 43.81096319139264, 'rmse': 225.6434108753146}
Testing for ensemble indices from 0 to 2
Best test metrics: {'mae': 43.7425880649186, 'rmse': 225.58439066425086}
Testing for ensemble indices from 0 to 3
Best test metrics: {'mae': 43.88726677477908, 'rmse': 225.7364938271056}
Testing for ensemble indices from 0 to 4
Best test metrics: {'mae': 43.81713467290032, 'rmse': 225.52863490139404}
Testing for ensemble indices from 0 to 5
Best test metrics: {'mae': 43.63538441082656, 'rmse': 225.1818673367202}
Testing for ensemble indices from 0 to 6
Best test metrics: {'mae': 43.5354812240608, 'rmse': 225.0400727201736}
Testing for ensemble indices from 0 to 7
Best test metrics: {'mae': 43.52160822103441, 'rmse': 224.98791414213858}
Testing for ensemble indices from 0 to 8
Best test metrics: {'mae': 43.47520143844204, 'rmse': 224.9298281182478}
Testing for ensemble indices from 0 to 9
Best test metrics: {'mae': 43.39384307836924, 'rmse': 224.76270912629127}
```

**Comment by weihua916** [comment]
`rel-trial-site` 0.370 MAE -> 0.341 MAE
```
Testing for ensemble indices from 0 to 0
Best test metrics: {'mae': 0.37000128850641656, 'rmse': 0.5661986556030657}
Testing for ensemble indices from 0 to 1
Best test metrics: {'mae': 0.3710682810301148, 'rmse': 0.5475061120784728}
Testing for ensemble indices from 0 to 2
Best test metrics: {'mae': 0.36860209218085593, 'rmse': 0.5290578931433301}
Testing for ensemble indices from 0 to 3
Best test metrics: {'mae': 0.35709842442414585, 'rmse': 0.5098868264818205}
Testing for ensemble indices from 0 to 4
Best Val metrics: {'mae': 0.3985401766198745, 'rmse': 0.540204892624096}
Best test metrics: {'mae': 0.34794154929974547, 'rmse': 0.5006839801508399}
Testing for ensemble indices from 0 to 5
Best test metrics: {'mae': 0.34552202869116566, 'rmse': 0.4983915341639107}
Testing for ensemble indices from 0 to 6
Best test metrics: {'mae': 0.34204072530825635, 'rmse': 0.4965095267908429}
Testing for ensemble indices from 0 to 7
Best test metrics: {'mae': 0.3449121169209921, 'rmse': 0.49949541903744965}
Testing for ensemble indices from 0 to 8
Best test metrics: {'mae': 0.3424676268067908, 'rmse': 0.4976470493495589}
Testing for ensemble indices from 0 to 9
Best test metrics: {'mae': 0.3414995579461586, 'rmse': 0.49436677265546464}
```

**Comment by weihua916** [comment]
rel-trial-withdrawal 0.6674 AUPRC -> 0.6793 AUPRC

```
Testing for ensemble indices from 0 to 0
Best test metrics: {'multilabel_auprc_micro': 0.6674279415354915, 'multilabel_auprc_macro': 0.3430230854490475, 'm
ultilabel_auroc_micro': 0.9071878312094048, 'multilabel_auroc_macro': 0.7880748365029719}
Testing for ensemble indices from 0 to 1
Best test metrics: {'multilabel_auprc_micro': 0.6789159730397097, 'multilabel_auprc_macro': 0.35032852764579053, '
multilabel_auroc_micro': 0.9101562913303353, 'multilabel_auroc_macro': 0.7975997777692739}
Testing for ensemble indices from 0 to 2
Best test metrics: {'multilabel_auprc_micro': 0.6764744915439789, 'multilabel_auprc_macro': 0.3502540825301176, 'm
ultilabel_auroc_micro': 0.9100100546463943, 'multilabel_auroc_macro': 0.8006199173333666}
Testing for ensemble indices from 0 to 3
Best test metrics: {'multilabel_auprc_micro': 0.6788432780258908, 'multilabel_auprc_macro': 0.35068203937367143, '
multilabel_auroc_micro': 0.9106288623016047, 'multilabel_auroc_macro': 0.8009634686604316}
Testing for ensemble indices from 0 to 4
Best test metrics: {'multilabel_auprc_micro': 0.6817943015552497, 'multilabel_auprc_macro': 0.3502062815350946, 'm
ultilabel_auroc_micro': 0.9112697741163195, 'multilabel_auroc_macro': 0.8013515456889084}
Testing for ensemble indices from 0 to 5
Best test metrics: {'multilabel_auprc_micro': 0.6807926288151167, 'multilabel_auprc_macro': 0.34876673597774926, '
multilabel_auroc_micro': 0.9107413365140663, 'multilabel_auroc_macro': 0.8003708809722503}
Testing for ensemble indices from 0 to 6
Best test metrics: {'multilabel_auprc_micro': 0.6810373780633608, 'multilabel_auprc_macro': 0.3522999125406001, 'm
ultilabel_auroc_micro': 0.9108999517345886, 'multilabel_auroc_macro': 0.8001672844932975}
Testing for ensemble indices from 0 to 7
Best test metrics: {'multilabel_auprc_micro': 0.6805384364481191, 'multilabel_auprc_macro': 0.348209430968972, 'mu
ltilabel_auroc_micro': 0.9106294737152995, 'multilabel_auroc_macro': 0.7991523604865165}
Testing for ensemble indices from 0 to 8
Best test metrics: {'multilabel_auprc_micro': 0.6798769932912736, 'multilabel_auprc_macro': 0.35025302396245667, 'multilabel_auroc_micro': 0.9105611145553507, 'multilabel_auroc_macro': 0.7992348955678269}
Testing for ensemble indices from 0 to 9
Best test metrics: {'multilabel_auprc_micro': 0.6793587262841474, 'multilabel_auprc_macro': 0.3467477424207713, 'multilabel_auroc_micro': 0.910323937856534, 'multilabel_auroc_macro': 0.7981673154471447}
```

File: examples/gnn_node.py [hunk 1]
```diff
@@ -168,25 +166,44 @@ def test(loader: NeighborLoader) -> np.ndarray:
...
-state_dict = None
-best_val_metric = 0 if higher_is_better else math.inf
-for epoch in range(1, args.epochs + 1):
-    train_loss = train()
+state_dicts = []
+
+for ensemble_idx in range(args.num_ensembles):
+    print(f"===Training for {ensemble_idx}-th ensemble index.")
+    model.reset_parameters()
+    optimizer = torch.optim.Adam(model.parameters(), lr=args.lr)
```
Comment (rusty1s) [inline examples/gnn_node.py:174]: personally would prefer to just instantiate new models. Allows you also to do a hyperparam sweep if you want.

**Review by rusty1s — APPROVED** [review]

File: examples/gnn_node.py (hunk 1 above)
Comment (weihua916) [reply examples/gnn_node.py:174]: Sounds good.

**Pull request merged**